src/bijectors/leaky_relu.jl

```diff
@@ -0,0 +1,95 @@
+"""
+    LeakyReLU{T, N}(α::T) <: Bijector{N}
+
+Defines the invertible mapping
+
+    x ↦ x if x ≥ 0 else αx
+
+where α > 0.
+"""
+struct LeakyReLU{T, N} <: Bijector{N}
+    α::T
+end
+
+LeakyReLU(α::T; dim::Val{N} = Val(0)) where {T<:Real, N} = LeakyReLU{T, N}(α)
+LeakyReLU(α::T; dim::Val{N} = Val(D)) where {D, T<:AbstractArray{<:Real, D}, N} = LeakyReLU{T, N}(α)
+
+up1(b::LeakyReLU{T, N}) where {T, N} = LeakyReLU{T, N + 1}(b.α)
+
+# (N=0) Univariate case
+function (b::LeakyReLU{<:Any, 0})(x::Real)
+    mask = x < zero(x)
+    return mask * b.α * x + (1 - mask) * x
+end
+(b::LeakyReLU{<:Any, 0})(x::AbstractVector{<:Real}) = b.(x)
+
+function (ib::Inverse{<:LeakyReLU, 0})(y::Real)
+    mask = y < zero(y)
+    return mask * (y / ib.orig.α) + (1 - mask) * y
+end
+(ib::Inverse{<:LeakyReLU{<:Any}, 0})(y::AbstractVector{<:Real}) = ib.(y)
+
+function logabsdetjac(b::LeakyReLU{<:Any, 0}, x::Real)
+    mask = x < zero(x)
+    J = mask * b.α + (1 - mask) * one(x)
+    return log(abs(J))
+end
+logabsdetjac(b::LeakyReLU{<:Real, 0}, x::AbstractVector{<:Real}) = logabsdetjac.(b, x)
+
+
+# We implement `forward` by hand since we can re-use the computation of
+# the Jacobian of the transformation. This will lead to faster sampling
+# when using `rand` on a `TransformedDistribution` making use of `LeakyReLU`.
+function forward(b::LeakyReLU{<:Any, 0}, x::Real)
+    mask = x < zero(x)
+    J = mask * b.α + (1 - mask) * one(x)
+    return (rv=J * x, logabsdetjac=log(abs(J)))
+end
+
+# Batched version
+function forward(b::LeakyReLU{<:Any, 0}, x::AbstractVector)
+    mask = x .< zero(eltype(x))
+    J = mask .* b.α .+ (1 .- mask) .* one(eltype(x))
```

Comment:
Maybe
```suggestion
    mask = x .< zero(eltype(x))
    J = @. (x < zero(x)) * b.α + (x > zero(x)) * one(x)
```
would be more efficient (since it avoids the allocation of `mask`) and more stable (since it does not require computations based on types).

As a side remark, IMO it is a bit unfortunate that currently so many almost identical implementations of a function are needed to define a bijector. Maybe this could be resolved by defining defaults on a high level for basically all bijectors (or a subtype of BaseBijectors, similar to BaseKernels in KernelFunctions) that call just in one or two methods that users would have to implement if their bijectors belong to these simple standard groups. With Julia >= 1.3 it is also possible to define methods for abstract functions such as
```julia
(f::Bijector{T,0})(x::AbstractVecOrMat{<:Real}) where T = map(f, x)
```

Reply:
> would be more efficient (since it avoids the allocation of mask) and more stable (since it does not require computations based on types).

Won't this call `zero(x)` n times though, i.e. have the same allocation? Other than that, it seems like a good idea :+1: 

> As a side remark

100%. 

> Maybe this could be resolved by defining defaults on a high level for basically all bijectors (or a subtype of BaseBijectors, similar to BaseKernels in KernelFunctions) that call just in one or two methods that users would have to implement if their bijectors belong to these simple standard groups.

I'm unaware of what KernelFunctions.jl does. Are they simply not making the `struct` callable? If so, it was a deliberate design-decision to go with making structs callable when we started out. We were debating whether or not to do this or include some `transform` method so we could define "one method to rule them all" on an abstract type. Ended up going with the "callable struct" approach, but it def has it's issues, i.e. redundant code. 

I've recently played around a bit with actually using a `transform` method under the hood, but adding a macro which allows you to say `@deftransform function transform(b, x) ... end` and we just add a `(b::Bijector)(x) = transform(b, x)` just after the method declaration. This would also allow us to implement in-place versions of all the methods, i.e. `transform!(b, x, out)`, `logabsdetjac(b, x, out)`, and so on. Thoughts?

> With Julia >= 1.3 it is also possible to define methods for abstract functions such as

Woah, really? If so that would solve the problem, no?

Reply:
> I'm unaware of what KernelFunctions.jl does. Are they simply not making the struct callable? If so, it was a deliberate design-decision to go with making structs callable when we started out. We were debating whether or not to do this or include some transform method so we could define "one method to rule them all" on an abstract type. Ended up going with the "callable struct" approach, but it def has it's issues, i.e. redundant code.

No, KernelFunctions API is based on callable structs only (using a function was removed a while ago). But e.g. translation-invariant kernels are built all in the same way, usually they use some simple function to evaluate the distance between the inputs (e.g. using Distances) and then they apply some nonlinear mapping afterwards. With this special structure also all kind of optimizations are possible when constructing kernel matrices etc. Hence there is a special type of such kernels (SimpleKernel IIRC), and then users just define their metric and the nonlinear mapping and get everything else for free. There's more information here: https://juliagaussianprocesses.github.io/KernelFunctions.jl/dev/create_kernel/

Reply:
To add on to the above, because it sucks to have to implement `(b::Bijector)`, `logabsdetjac` and then finally `forward`, would it be an idea to add a macro that would allow you to define all in one go? 

I.e. something like the following. IMHO it feels a bit like "overkill" (plus I bet the macro below is not the cleanest way of doing it), but curious what you think. 
```julia
"""
    @bijector function forward(b::Bijector, x) ... end

Takes the method `forward` and uses it to define both `transform`
and `logabsdetjac`, while ensuring that any shared computation is
taken advantage of in `forward`.
"""
macro bijector(expr)
    def = MacroTools.splitdef(expr)
    body = def[:body]
    args = def[:args]

    # Figure out what is shared, and what isn't
    shared_exprs = [] # beginning of `forward`
    transform_exprs = []
    logjac_exprs = []
    tail_exprs = [] # goes at the end of `forward`

    for expr in body.args
        # When using `rv = ...` and `logabsdetjac = ...`
        if Meta.isexpr(expr, :(=)) && expr.args[1] == :rv
            push!(transform_exprs, Expr(:return, expr.args[2]))
            push!(tail_exprs, expr)
        elseif Meta.isexpr(expr, :(=)) && expr.args[1] == :logabsdetjac
            push!(logjac_exprs, Expr(:return, expr.args[2]))
            push!(tail_exprs, expr)
        else
            # If still sharing, add those expressions
            issharing = (length(transform_exprs) == 0) && (length(logjac_exprs) == 0)
            if issharing
                push!(shared_exprs, expr)
            else
                push!(tail_exprs, expr)
            end
        end
    end

    # Add the shared computation
    transform_full_exprs = copy(shared_exprs)
    append!(transform_full_exprs, transform_exprs)

    logjac_full_exprs = copy(shared_exprs)
    append!(logjac_full_exprs, logjac_exprs)

    # Remove the redundant macro's from the `forward` body
    forward_full_exprs = vcat(shared_exprs, tail_exprs)

    return quote
        $(Bijectors).@deftransform function $(Bijectors).transform($(args...))
            $(transform_full_exprs...)
        end

        function $(Bijectors).logabsdetjac($(args...))
            $(logjac_full_exprs...)
        end

        function $(Bijectors).forward($(args...))
            $(forward_full_exprs...)
        end
    end
end
```
in which case something like
```julia
@macroexpand Bijectors.@bijector function forward(b::Bijectors.Exp{0}, x::Real)
    # Shared computation
    z = exp(x)
    
    rv = begin
        # `(b::Bijector)(x)` goes here
        z
    end
    logabsdetjac = begin
        # `logabsdetjac(b, x)` goes here
        log(z)
    end

    return (rv = rv, logabsdetjac = logabsdetjac)
end
```
is expanded to
```julia
$(Expr(:meta, :doc))
function (Bijectors).transform(var"#4#b"::(Bijectors.Bijectors).Exp{0}, var"#7#x"::Bijectors.Real)
    var"#5#z" = Bijectors.exp(var"#7#x")
    return var"#5#z"
end

(var"#4#b"::(Bijectors.Bijectors).Exp{0})(var"#9#x"::Bijectors.Real) = begin
    (Bijectors).transform(var"#4#b", var"#9#x")
end

function (Bijectors).logabsdetjac(var"#11#b"::(Bijectors.Bijectors).Exp{0}, var"#12#x"::Bijectors.Real)
    var"#10#z" = Bijectors.exp(var"#12#x")
    return Bijectors.log(var"#10#z")
end
function (Bijectors).forward(var"#16#b"::(Bijectors.Bijectors).Exp{0}, var"#17#x"::Bijectors.Real)
    var"#13#z" = Bijectors.exp(var"#17#x")
    var"#14#rv" = var"#13#z"
    var"#15#logabsdetjac" = Bijectors.log(var"#13#z")
    return (rv = var"#14#rv", logabsdetjac = var"#15#logabsdetjac")
end
```

Reply:
> But e.g. translation-invariant kernels are built all in the same way, usually they use some simple function to evaluate the distance between the inputs (e.g. using Distances) and then they apply some nonlinear mapping afterwards.

Ah, I see. I don't think it's so easy to do for bijectors as we have much more structure than just "forward" evaluation, no?